src/net/dns.cc

```diff
@@ -856,6 +857,11 @@ class net::dns_resolver::impl
                         e.udp.f = make_ready_future<>();
                         return -1;
                     }
+
+                    // ensure that no exception from channel.send is left uncaught
+                    e.udp.f = e.udp.f.handle_exception_type([](std::system_error const& e){
+                        dns_log.warn("UDP send exception: {}", e.what());
```

Comment:
Better to do this after the sendv(). Change finally to then_wrapped so you can handle it there. Fewer extra continuations, and more local error handling.

@elcallio please review too.

Reply:
But in that case, I have to remove the fast-fail branch on line 849, right?

Reply:
It sorta looks like you can move the fast-fail case directly into the last `finally` (but switched to then_wrapped) without sacrificing the optimization. It's either going to be available at that point (finally is just a fancy then_wrapped) or you'd be waiting on it to become available (the case for the code you are adding now).

but yeh i didn't look in complete detail at what's going on.

Reply:
@avikivity `then_wrapped` solution will not be able to fast-fail (because we have no notion of fast/continued there). If you want to minimize continuations, why not just wrap with another/same check, i.e. `f.available()`, if, and failed -> fast-fail, if success -> success, if not available(), add this handle_exception. Since we will fast-succeed in 90% of cases anyway, it should not add much overhead.

Reply:
I'm trying the then_wrapped solution, but it gets somewhat awkward to make it work with the fast fail, in contrast, I think @elcallio suggestion is simpler to get right

Reply:
@elcallio i'm a little curious about this statement, but perhaps i'm about to learn something new!

> then_wrapped solution will not be able to fast-fail (because we have no notion of fast/continued there)

in the follow proposed change, if `e.udp.f` is available, then it implies that the full chain above the available check, including the finally blocks were fully evaluated (otherwise, how would it be available)?

```c++
                    e.udp.f = e.udp.f.finally([&e, p = std::move(p)]() mutable {
                        return e.udp.channel.send(e.udp.dst, std::move(p));;
                    }).finally([fd, me = shared_from_this()] {
                        me->release(fd);
                    });
                    if (e.udp.f.available()) {
                        // if we have a fast-fail, give error.
                        if (e.udp.f.failed()) {
```

if that's the case, then checking inside an attached `then_wrapped` seems like it would be equivalent in cost (because a `then_wrapped` is already there in the form of `finally`).

Reply:
@dotnwat The problem I encountered is that fast-fail needs to see the exception, but if then_wrapped absorbs the exception, we lose that data channel. I tried to change from future<> to future<std::error_code>, but the change would be bigger and It would have to deal with the future destruction done by me->release(fd)

Reply:
Thanks @elcallio for confirming my understanding. You're right, of course about the return code! I realize now in the back of my mind I thinking about capturing a stack variable reference, but that's massively problematic :)